tasks/Task.cpp

```diff
@@ -453,6 +461,147 @@ string Task::convertFrameModeToPixelFormat(frame_mode_t format, uint8_t data_dep
+        device.GetNodeMap()->GetNode("GevSCFTD");
+    pStreamChannelFrameTransmissionDelay->SetValue(scftd);
+
+    waitDevicePTPNegotiation(device, system);
+}
+
+void Task::waitDevicePTPNegotiation(Arena::IDevice& current_device,
+    Arena::ISystem& system)
+{
+    // Wait for devices to negotiate their PTP relationship
+    //    Before starting any PTP-dependent actions, it is important to wait for
+    //    the devices to complete their negotiation; otherwise, the devices may
+    //    not yet be synced. Depending on the initial PTP state of each camera,
+    //    it can take about 40 seconds for all devices to autonegotiate. Below,
+    //    we wait for the PTP status of each device until there is only one
+    //    'Master' and the rest are all 'Slaves'. During the negotiation phase,
+    //    multiple devices may initially come up as Master so we will wait until
+    //    the ptp negotiation completes.
+
+    auto current_device_ip =
+        Arena::GetNodeValue<int64_t>(current_device.GetNodeMap(), "CurrentIPAddress");
+    system.UpdateDevices(100);
+    vector<Arena::DeviceInfo> device_infos = system.GetDevices();
+
+    auto start_time = base::Time::now();
+    while (true) {
+        bool masterFound = false;
+        bool restartSyncCheck = false;
+
+        // check devices
+        for (auto device_info : device_infos) {
+            if (device_infos.size() !=
+                static_cast<size_t>(_camera_config.get().ptp_sync.number_of_cameras)) {
+                LOG_WARN_S << "Not all cameras are discovered yet." << endl;
+                system.UpdateDevices(100);
+                device_infos.clear();
+                device_infos = system.GetDevices();
+            }
+            GenICam::gcstring ptp_status;
+            if (device_info.IpAddress() == current_device_ip) {
+                // get PTP status
+                ptp_status =
+                    Arena::GetNodeValue<GenICam::gcstring>(current_device.GetNodeMap(),
+                        "PtpStatus");
+            }
+            else {
+                ArenaDevice dev(system.CreateDevice(device_info), system);
+                ptp_status =
+                    Arena::GetNodeValue<GenICam::gcstring>(dev.device->GetNodeMap(),
+                        "PtpStatus");
+            }
+
+            if (ptp_status == "Master") {
+                if (masterFound) {
+                    // Multiple masters -- ptp negotiation is not complete
+                    restartSyncCheck = true;
+                    break;
+                }
+
+                masterFound = true;
+            }
+            else if (ptp_status != "Slave") {
+                // Uncalibrated state -- ptp negotiation is not complete
+                restartSyncCheck = true;
+                break;
+            }
+        }
+
+        // A single master was found and all remaining cameras are slaves
+        if (!restartSyncCheck && masterFound)
+            break;
```

Comment:
Create a separate isPTPSyncSettled(device_info) method  that returns true or false and then

```
if (isPTPSyncSettled(device_info)) {
   return;
}
```